c-dependencies/js-compute-runtime/js-compute-builtins.cpp

```diff
@@ -3238,6 +3282,583 @@ namespace FetchEvent {
+    }
+  }
+
+  JSUrlSearchParams* get_params(JSObject* self) {
+    return (JSUrlSearchParams*)JS::GetReservedSlot(self, Slots::Params).toPrivate();
+  }
+
+  bool check_receiver(JSContext* cx, HandleObject self, const char* method_name);
+  JSObject* create(JSContext* cx, HandleValue params_val);
+
+  const unsigned ctor_length = 1;
+
+  bool constructor(JSContext* cx, unsigned argc, Value* vp) {
+    CallArgs args = CallArgsFromVp(argc, vp);
+
+    RootedObject self(cx, create(cx, args.get(0)));
+    if (!self) return false;
+
+    args.rval().setObject(*self);
+    return true;
+  }
+
+  bool append(JSContext* cx, unsigned argc, Value* vp) {
+    METHOD_HEADER(2)
+    if (!detail::append(cx, self, args[0], args[1], "append"))
+      return false;
+
+    args.rval().setUndefined();
+    return true;
+  }
+
+  bool delete_(JSContext* cx, unsigned argc, Value* vp) {
+    METHOD_HEADER_WITH_NAME(1, "delete")
+    const auto params = (JSUrlSearchParams*)JS::GetReservedSlot(self, Slots::Params).toPrivate();
+
+    auto name = encode(cx, args.get(0));
+    if (!name.data) return false;
+
+    jsurl::params_delete(params, &name);
+    args.rval().setUndefined();
+    return true;
+  }
+
+  bool has(JSContext* cx, unsigned argc, Value* vp) {
+    METHOD_HEADER(1)
+    const auto params = (JSUrlSearchParams*)JS::GetReservedSlot(self, Slots::Params).toPrivate();
+
+    auto name = encode(cx, args.get(0));
+    if (!name.data) return false;
+
+    args.rval().setBoolean(jsurl::params_has(params, &name));
+    return true;
+  }
+
+  bool get(JSContext* cx, unsigned argc, Value* vp) {
+    METHOD_HEADER(1)
+    const auto params = (JSUrlSearchParams*)JS::GetReservedSlot(self, Slots::Params).toPrivate();
+
+    auto name = encode(cx, args.get(0));
+    if (!name.data) return false;
+
+    const SpecSlice slice = jsurl::params_get(params, &name);
+    if (!slice.data) {
+      args.rval().setNull();
+      return true;
+    }
+
+    RootedString str(cx, JS_NewStringCopyUTF8N(cx, JS::UTF8Chars((char*)slice.data, slice.len)));
+    if (!str) return false;
+    args.rval().setString(str);
+    return true;
+  }
+
+  bool getAll(JSContext* cx, unsigned argc, Value* vp) {
+    METHOD_HEADER(1)
+    const auto params = (JSUrlSearchParams*)JS::GetReservedSlot(self, Slots::Params).toPrivate();
+
+    auto name = encode(cx, args.get(0));
+    if (!name.data) return false;
+
+    const jsurl::CVec<SpecSlice> values = jsurl::params_get_all(params, &name);
+
+    RootedObject result(cx, JS::NewArrayObject(cx, values.len));
+    if (!result) return false;
+
+    RootedString str(cx);
+    RootedValue str_val(cx);
+    for (size_t i = 0; i < values.len; i++) {
+      const SpecSlice value = values.ptr[i];
+      str = JS_NewStringCopyUTF8N(cx, JS::UTF8Chars((char*)value.data, value.len));
+      if (!str) return false;
+
+      str_val.setString(str);
+      if (!JS_SetElement(cx, result, i, str_val))
+        return false;
+    }
+
+    free(values.ptr);
```

Comment:
Do we have a guarantee that Rust-on-wasm uses wasi-libc's malloc/free? I suppose we'd have much deeper problems if that weren't the case, but this asymmetry still bothers me a tiny bit; I'd prefer for the `values` to be passed back into a Rust FFI call to be freed on that side.

Reply:
For this particular case, I was able to avoid this by just adding a `Drop` impl for `CVec`—which is the right thing to do anyway.

However, on a somewhat fundamental level we really can't avoid this: we have to pass strings to Rust that were created by SpiderMonkey (I think in all cases via the `encode` function). I guess we could in theory never have the Rust code take ownership of them and always copy, but that seems pretty unfortunate.

I also think this should be something we can indeed rely on: SpiderMonkey contains lots of Rust code precisely for string handling, so I would strongly assume that the build config sorts out all the allocator stuff :)

Reply:
Fair enough!